pkg/pkgrepository/app_template.go

```diff
@@ -18,30 +27,116 @@ func (a *App) template(dirPath string) exec.CmdRunResult {
+	result, _ = template.TemplateDir(dirPath)
+	if result.Error != nil {
+		return result
+	}
 
-	for _, tpl := range a.app.Spec.Template {
-		var template ctltpl.Template
+	// second templating applies a bunch of overlays and filters, some of which could be migrated to go code
+	template = a.templateFactory.NewYtt(*(a.app.Spec.Template[1]).Ytt, genericOpts)
+	result = template.TemplateStream(strings.NewReader(result.Stdout), dirPath)
+	if result.Error != nil {
+		return result
+	}
 
-		switch {
-		case tpl.Ytt != nil:
-			template = a.templateFactory.NewYtt(*tpl.Ytt, genericOpts)
-		case tpl.Kbld != nil:
-			template = a.templateFactory.NewKbld(*tpl.Kbld, genericOpts)
-		default:
-			return exec.NewCmdRunResultWithErr(fmt.Errorf("Unsupported way to template"))
+	// intermediate phase deserializes and reserializes all of the resources
+	// which guarantees that they only include fields this version of kc knows about.
+	streamForSomeReason := strings.NewReader(result.Stdout)
+	buf := make([]byte, streamForSomeReason.Len())
+	_, err := io.ReadFull(streamForSomeReason, buf)
+	if err != nil {
+		result.Error = err
+		return result
+	}
+	resources, err := FilterResources(buf)
+	if err != nil {
+		result.Error = err
+		return result
+	}
+
+	// third templating inserts a kapp rebase rule to allow noops on identical resources provided by multiple pkgrs
+	template = a.templateFactory.NewYtt(*(a.app.Spec.Template[2]).Ytt, genericOpts)
+	result = template.TemplateStream(strings.NewReader(resources), dirPath)
+
+	a.log.Info("returning from the great yaml caper")
+	return result
+}
+
+// FilterResources takes a multi-doc yaml byte array of the templated
+// contents of a PKGR, and filters each resource by deserializing and re-serializing
+func FilterResources(yamls []byte) (string, error) {
+	sch := runtime.NewScheme()
+	err := scheme.AddToScheme(sch)
+	if err != nil {
+		return "", err
+	}
+	err = datapackagingv1alpha1.AddToScheme(sch)
+	if err != nil {
+		return "", err
+	}
+	decoder := serializer.NewCodecFactory(sch).UniversalDeserializer().Decode
+
+	filteredYamls := []string{}
+	docs, err := yamlDocs(yamls)
+	if err != nil {
+		return "", err
+	}
+	for _, resourceYAML := range docs {
+
+		obj, gKV, err := decoder(resourceYAML, nil, nil)
+		if err != nil {
+			return "", err
+		}
+		kind := gKV.Kind
+		if gKV.Group != "data.packaging.carvel.dev" {
+			return "", fmt.Errorf("got unexpected group: %s", gKV.Group)
 		}
 
-		if isStream {
-			result = template.TemplateStream(strings.NewReader(result.Stdout), dirPath)
-		} else {
-			result, isStream = template.TemplateDir(dirPath)
+		if gKV.Version != "v1alpha1" {
+			return "", fmt.Errorf("got unexpected version: %s", gKV.Version)
 		}
-		if result.Error != nil {
+
+		switch kind {
+		case "Package":
+			p := obj.(*datapackagingv1alpha1.Package)
+			yp := printers.YAMLPrinter{}
+			buffy := new(bytes.Buffer)
+			yp.PrintObj(p, buffy)
```

Comment:
i think Print is for "pretty printing". we should just be able to use yaml.Marshal() (https://github.com/kubernetes-sigs/yaml/blob/master/yaml.go#L31)